### PR TITLE
Make `ToPickle` a `Generic`

### PR DESCRIPTION
**Pull request by hendrikmakait**
Extracted from #8096

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/protocol/serialize.py [hunk 1]
```diff
@@ -572,19 +574,21 @@ class ToPickle:
     to False, the scheduler will raise an exception instead.
     """
 
-    def __init__(self, data):
+    data: T
+
+    def __init__(self, data: T) -> None:
```
Comment (crusaderky) [inline distributed/protocol/serialize.py:579]:
```suggestion
    def __init__(self, data: T):
```
Return type of `__init__` should always be implicit

File: distributed/protocol/serialize.py [hunk 2]
```diff
@@ -572,19 +574,21 @@ class ToPickle:
...
+    def __repr__(self) -> str:
         return "<ToPickle: %s>" % str(self.data)
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
         return isinstance(other, type(self)) and other.data == self.data
 
-    def __ne__(self, other):
+    def __ne__(self, other: object) -> bool:
         return not (self == other)
```
Comment (crusaderky) [inline distributed/protocol/serialize.py:589]: out of scope: fairly sure this method is redundant with `__eq__`

Comment (hendrikmakait) [reply distributed/protocol/serialize.py]: Good point, didn't even pay attention to this. Dropped `__ne__` as a drive-by fix.

**Review by crusaderky — APPROVED** [review]
Good to merge as soon as CI finishes

**Pull request merged**